File: cpp/cor-error/lib/bktree.cc
#include <string>
#include <unordered_map>
#include <vector>
#include <cstring>
#include <queue>
#include <stack>
#include <algorithm>
#include <stdio.h> 
#include <fstream>
#include <numeric>
#include <iostream>
#include <cmath>

#include "bktree.h"
#include "utils.h"

namespace BkTree {
    bool sort_by_score(const std::pair<float, std::wstring> &a, const std::pair<float, std::wstring> &b) {
        return (a.first > b.first);
    } 

    std::wstring node::get_item() const {
        return item;
    }

    void node::set_item(std::wstring& _item) {
        item = _item;
    }

    float node::get_hot() const {
        return hot;
    }

    void node::set_hot(const float &h) {
        hot = h;
    }

    int tree::edit_distance(const std::wstring& wstr1, const std::wstring& wstr2) {
        int m = wstr1.size(), n = wstr2.size(), dp[m + 1], i, j, prev, tmp;
        for (i = 0; i <= m; i++) dp[i] = i;
        for (i = 1; i <= n; i++) {
            prev = i-1;
            dp[0] = i;
            for (j = 1; j <= m; j++) {
                tmp = dp[j];
                if (wstr1[j - 1] == wstr2[i - 1]) dp[j] = prev;
                else dp[j] = std::min(std::min(dp[j], prev), dp[j - 1]) + 1;
                prev = tmp;
            }
        }

        return dp[m];          
    }

    tree::~tree() {
        tree::destroy(root);
    }

    void tree::destroy(node* n) {
        if (n->get_item().empty()) return;

        n->get_item().clear();
        for (auto c:n->children) tree::destroy(c.second);
    }

    void tree::add(std::string str, float hot) {
        node *n = root;
        std::wstring wstr = utils::str2wstr(str);
        if (root->get_item().empty()) {
            root->set_item(wstr); 
            root->set_hot(hot);
        }
        else while (1) {
            int d = edit_distance(n->get_item(), wstr);
            if (d == 0) break;
            auto search = n->children.find(d);
            if (search != n->children.end()) n = search->second;
            else {
                n->children[d] = new node(wstr, hot);
                break;
            }
        }
    }

    std::vector<std::pair<float, std::wstring>> tree::search(std::string item, int value) {
        std::vector<std::pair<float, std::wstring>> res;
        if (root->get_item().empty()) return res;
        int d, idx;
        node *n;
        std::queue<node *> q;
        q.push(root);
        std::wstring wstr = utils::str2wstr(item);
        float score = 0, _score;
        while (q.size()) {
            n = q.front();
            q.pop();
            
            d = edit_distance(n->get_item(), wstr);
            _score = pow(pow_count, d)*n->get_hot();
            if (d <= value && score < _score) {
                score = _score;
                res.push_back(make_pair(_score, n->get_item()));
            }

            for (idx = std::max(1, d-value); idx <= d+value; idx++) if (n->children.count(idx)) if (pow_count*n->children[idx]->get_hot() > score) q.push(n->children[idx]);
        }

        if (res.size()) std::sort(res.begin(), res.end(), sort_by_score); 

        return res;        
    }  

    int tree::levenshtein_distance(const std::string &str1, const std::string &str2) {
        std::wstring wstr1 = utils::str2wstr(str1), wstr2 = utils::str2wstr(str2);
        return edit_distance(wstr1, wstr2);
    }

    int tree::longest_common_seq(const std::string &str1, const std::string &str2) {
        int len1 = str1.size(), len2 = str2.size(), dp[len1 + 1], i, j;
        for (i = 0; i <= len1; i++) dp[i] = 0;
        for (j = 1; j <= len2; j++) for (i = 1; i <= len1; i++) {
            if (str1[i - 1] == str2[j - 1]) dp[i] = dp[i-1] + 1;
            else dp[i] = std::max(dp[i-1], dp[i]);
        }
        return dp[len1];           
    }
} // namespace BkTree

File: java/cor-error/BkTree.java
import java.util.*;

public class BkTree {
    private Node root;
    private double powCount, gatherThro, predThro;
    private HashMap<String, List<String>> pyDict;
    private int value;
    private Boolean useCustomFilter;
    private final CustomInterface customFilter;    

    public BkTree(double gatherThro, double predThro, int value, CustomInterface customFilter) {
        this.root = new Node();
        this.powCount = 0.1;
        this.value = value;
        this.pyDict = new HashMap<>();
        this.gatherThro = gatherThro;
        this.predThro = predThro;    
        this.useCustomFilter = true;
        this.customFilter = customFilter;
    }

    public BkTree(double gatherThro, double predThro, int value) {
        this.root = new Node();
        this.powCount = 0.1;
        this.value = value;
        this.pyDict = new HashMap<>();
        this.gatherThro = gatherThro;
        this.predThro = predThro;
        this.useCustomFilter = false;
        this.customFilter = null;        
    }

    public Node getRoot() {
        return this.root;
    }

    public static String correct(String name, Collection<String> expectNames) {
        return correct(name,expectNames,0.75,0.7,2);
    }

    public static String correct(String name, Collection<String> expectNames, double gatherThro, double predThro, int distance) {
        BkTree tree = new BkTree(gatherThro, predThro, distance);
        List<String> pyName = Arrays.asList(Pinyin.toPinyin(name, ",").split(","));
        List<String> newExpectNames = new ArrayList<>();
        List<List<String>> expectPyNames = new ArrayList<>();
        expectNames
                .forEach(expectName -> {
                    newExpectNames.add(expectName);
                    expectPyNames.add(Arrays.asList(Pinyin.toPinyin(expectName, ",").split(",")));
                });
        return tree.divine(name, pyName, newExpectNames, expectPyNames);
    }

    public static String correct(String name, List<String> pyName, List<String> names, List<List<String>> pyNames){
        BkTree tree = new BkTree(0.75, 0.7, 2);
        return tree.divine(name, pyName, names, pyNames);
    }

    public String divine(String name, List<String> pyName, List<String> names, List<List<String>> pyNames) {
        assert names.size() == pyNames.size();
        this.pyDict.clear();

        List<Pair<Integer, String>> samePyNames = new ArrayList<>();
        int idx;

        for (idx = 0; idx < names.size(); idx++) {
            if (name.equals(names.get(idx))) {
                return name;
            }

            if (pyNames.get(idx).equals(pyName)) {
                samePyNames.add(new Pair<>(editDistance(names.get(idx), name), names.get(idx)));
                continue;
            }

            this.pyDict.put(names.get(idx), pyNames.get(idx));
            add(names.get(idx), 1.0);
        }

        if (!samePyNames.isEmpty()) {
            samePyNames.sort(Comparator.comparingInt(Pair::getKey));
            return samePyNames.get(0).getValue();
        }

        List<Pair<Double, String>> res = search(name, this.value);
        restoreRoot();
        if (res.isEmpty()) {
            return null;
        }

        res.sort((o1, o2) -> o1.getKey() - o2.getKey() > 0.0 ? -1:1);

        double distance = res.get(0).getKey(), bestTgtThro = 0.0, bestSrcThro = 0.0;
        String pyname = String.join("", pyName), theName = null;

        for (idx = 0; idx < res.size(); idx++) {
            if (res.get(idx).getKey() < distance) {
                break;
            }
            String tName = res.get(idx).getValue();
            if (this.useCustomFilter && this.customFilter.Fillter(tName, name)) continue;
            String tNamePy = String.join("", this.pyDict.get(tName));
            double tgtThro = longestCommonSeq(tNamePy, pyname) * 1.0 / pyname.length();
            double srcThro = longestCommonSeq(tNamePy, pyname) * 1.0 / tNamePy.length();
            if (tgtThro > bestTgtThro) {
                bestTgtThro = tgtThro;
                theName = tName;
                bestSrcThro = srcThro;
            } else if (tgtThro == bestTgtThro && bestSrcThro < srcThro) {
                bestSrcThro = srcThro;
                theName = tName;
            }
        }

        if (bestTgtThro > this.gatherThro && bestSrcThro > this.predThro) {
            return theName;
        }
        return null;
    }

    private List<Pair<Double, String>> search(String word, int radius) {
        List<Pair<Double, String>> res = new ArrayList<>();
        if (this.root.getWord() == null) {
            return res;
        }

        int d, idx;
        double score = 0, tmpScore;
        Queue<Node> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            Node n = queue.poll();
            d = editDistance(n.getWord(), word);
            tmpScore = Math.pow(this.powCount, d)*n.getHot();
            if (d <= radius && score <= tmpScore) {
                score = tmpScore;
                res.add(new Pair<>(score, n.getWord()));
            }

            for (idx = Math.max(1, d-radius); idx <= d+radius; idx++) {
                if (n.getChildren().containsKey(idx) && Math.pow(this.powCount, d)*n.getChildren().get(idx).getHot() >= score) {
                    queue.offer(n.getChildren().get(idx));
                }
            }
        }
        return res;
    }

    public void restoreRoot() {
        destroy(this.root);
        this.root = new Node();
    }

    private void destroy(Node node) {
        if (node.getWord() != null) {
            node.setWord(null);
        }

        Iterator iter = node.getChildren().entrySet().iterator();
        while (iter.hasNext()) {
            Map.Entry entry = (Map.Entry) iter.next();
            destroy((Node) entry.getValue());
        }
        node.getChildren().clear();
    }

    public void add(String word, double hot) {
        Node n = this.root;
        if (root.getWord() == null) {
            root.setWord(word);
            root.setHot(hot);
        } else {
            while (true) {
                int d = editDistance(n.getWord(), word);
                if (d == 0) {
                    break;
                }
                if (n.getChildren().containsKey(d)) {
                    n = n.getChildren().get(d);
                } else {
                    n.getChildren().put(d, new Node(word, hot));
                    break;
                }
            }
        }
    }

    private int longestCommonSeq(String py1, String py2) {
        int m = py1.length(), n = py2.length(), prev;
        int[] dp = new int[m+1];
        for (int i = 0; i <= m; i++) {
            dp[i] = 0;
        }
        for (int j = 1; j <= n; j++) {
            for (int i = 1; i <= m; i++) {
                prev = dp[i-1];
                if (py1.charAt(i - 1) == py2.charAt(j - 1)) {
                    dp[i] = prev + 1;
                } else {
                    dp[i] = Math.max(dp[i], dp[i - 1]);
                }
            }
        }
        return dp[m];
    }

    private int editDistance(String word1, String word2) {
        int m = word1.length(), n = word2.length(), tmp, prev;
        int[] dp = new int[m + 1];
        for (int i = 0; i <= m; i++) {
            dp[i] = i;
        }
        for (int i = 1; i <= n; i++) {
            prev = i-1; dp[0] = i;
            for (int j = 1; j <= m; j++) {
                tmp = dp[j];
                if (word1.charAt(j - 1) == word2.charAt(i - 1)) {
                    dp[j] = prev;
                } else {
                    dp[j] = Math.min(Math.min(dp[j], dp[j - 1]), prev) + 1;
                }
                prev = tmp;
            }
        }
        return dp[m];
    }
}
